From f9aadff342a9a56d938ba47e300e341913e8e878 Mon Sep 17 00:00:00 2001
From: stakiar <stakiar@cloudera.com>
Date: Wed, 22 Feb 2017 10:32:54 -0800
Subject: [PATCH 1009/1164] CLOUDERA-BUILD: Fix for CDH-47856 given that
 HIVE-13705 was reverted

The commit modifies the Hive.mvFile(...) method so that if the destination
filesystem is local, an explicit fs.exists(destFilePath) is called before
executing the fs.rename(...) call.

Change-Id: I42486421f74a590ade1993486fe8ce42f8e7b867
---
 .../org/apache/hadoop/hive/ql/metadata/Hive.java   |    7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java b/ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java
index 9e3623f..ce4a925 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java
@@ -2707,7 +2707,12 @@ private static Path mvFile(HiveConf conf, FileSystem sourceFs, Path sourcePath,
        */
     
     int counter = 1;
-    if (!isRenameAllowed || isBlobStoragePath) {
+    // A call to fs.exists is necessary for the local filesystem due to CDH-47856 - while there is a fix for the reported
+    // bug upstream, that fix breaks CM metrics integration, so a custom fix has been added. The issue is that the standard
+    // local filesystem object allows rename-overwrites operations. However, HDFS does not, and this part of the commit logic
+    // requires that data be committed to a unique path (so that data is not overwritten).
+    final boolean isDestLocal = destFs.getUri().getScheme().equals("file") || destFs.getUri().getScheme().equals("pfile");
+    if (!isRenameAllowed || isBlobStoragePath ||isDestLocal) {
       while (destFs.exists(destFilePath)) {
         destFilePath =  new Path(destDirPath, name + ("_copy_" + counter) + (!type.isEmpty() ? "." + type : ""));
         counter++;
-- 
1.7.9.5

